### PR TITLE
feat: Added generic and SearchBar and Orama powered SearchBar.

### PR DESCRIPTION
**Pull request by ShogunPanda**
## Description

This PR introduces a Search Bar component and it also introduces a [Orama](https://oramasearch.com) based implementation of the same component.

### Check List

- [x] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [x] I have run `npx turbo format` to ensure the code follows the style guide.
- [x] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [x] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

**Comment by tniessen** [comment]
Is this a client-side index that is generated at build time? I am not familiar with Orama or Next.js but I believe @ovflowd suggested multiple alternatives and wanted to avoid client-side processing of search indices.

Refs: https://github.com/nodejs/admin/issues/766#issuecomment-1431560083

**Comment by ovflowd** [comment]
I also haven't checked this PR yet. But I asked the Orama team for something that allowed us leverage both world. A client-side only on first build but if Server-side is available to leverage that 🤔

Will check if this PR is implementing this 👀

**Comment by ShogunPanda** [comment]
This PR is opened to both possibilities.
At the moment I'm not generating any index and I'm just providing the component for storybook.
In the future the index can be generated on the server and downloaded via `fetch` in the `setup` call. Or can be generated in the browser (always in the `setup` call).

File: components/Common/OramaSearchBar/index.module.scss [hunk 1]
```diff
@@ -0,0 +1,24 @@
+.oramaSearchBarFooter {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.module.scss:1]: The style might need some changes for dark mode

image

File: components/Common/OramaSearchBar/index.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
+import { create, load, search } from '@orama/orama';
+import { useTheme } from 'next-themes';
+import { useCallback, useMemo, useState } from 'react';
+import { FormattedMessage } from 'react-intl';
+import { OramaLogoDark, OramaLogoLight } from './OramaLogo';
+import styles from './index.module.scss';
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx:6]: styles usually should be the last import 👀 

After all other imports (so before the `DocumentMapper` type

File: components/Common/OramaSearchBar/index.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import SearchBar from '../SearchBar';
+
+import type { Document, Orama, RawData, Schema } from '@orama/orama';
+import type { FC } from 'react';
+import type { SearchFunction, SearchResult } from '../../../types';
+
+type DocumentMapper = (doc: Document) => SearchResult;
+
+type FooterProps = {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx]: This can be a one-line type `type FooterProps = { ... }`

File: components/Common/OramaSearchBar/index.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  theme?: string;
+};
+
+type OramaSearchBarProps = {
+  schema: Schema;
+  index: RawData;
+  documentMapper?: DocumentMapper;
+};
+
+function identityDocumentMapper(doc: Document): SearchResult {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx]: We prefer arrow functions within the codebase.

File: components/Common/OramaSearchBar/index.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  mapper: DocumentMapper,
+  term: string
+): Promise<SearchResult[]> {
+  const results = await search(database, { term });
+
+  return results.hits.map(h => mapper(h.document));
+}
+
+const Footer: FC<FooterProps> = ({ theme }: FooterProps) => {
+  return (
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx:19]: No need of a return statement here

File: components/Common/OramaSearchBar/index.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    <div className={styles.oramaSearchBarFooter}>
+      <span className={styles.oramaSearchBarFooterLabel}>
+        <FormattedMessage id="components.oramaSearchBar.search.poweredBy" />
+      </span>
+      <a
+        href={'http://oramajs.io'}
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx]:
```suggestion
        href="http://oramajs.io"
```

File: components/Common/OramaSearchBar/index.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        target="_blank"
+        rel="noopener noreferrer"
+        className={styles.oramaSearchBarFooterLink}
+      >
+        {theme === 'dark' ? (
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx]: This can be memoized I guess. Also probably you want to use `useTheme` hook for the theme :)

File: components/Common/OramaSearchBar/index.tsx
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx:1]: I think all the application logic, like performing search and etc should be done in a Hook and utils files.

File: components/Common/OramaSearchBar/index.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  schema: Schema;
+  index: RawData;
+  documentMapper?: DocumentMapper;
+};
+
+function identityDocumentMapper(doc: Document): SearchResult {
+  return doc as unknown as SearchResult;
+}
+
+async function performSearch(
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx]: This should be on a dedicated place just for search logic, like an util or just a separate file within `OramaSearchBar`

File: components/Common/SearchBar/index.module.scss [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    .searchResultCode {
+      border-radius: 0.3rem;
+      font-weight: var(--font-weight-light);
+      padding: 0 6px;
+    }
+  }
+}
+
+[data-theme='light'] .searchResultCode {
```
Comment (ovflowd) [inline components/Common/SearchBar/index.module.scss:128]: I don't think we're using `data-theme` for themes anymore. Search for "dark" and "light"

**Review by ovflowd — COMMENTED** [review]
I haven't gone through the SearchBar and useKeypress hook yet, but this are my comments so far :)

**Comment by AugustinMauroy** [comment]
Hey!
You should update NewHeader with search feat.

**Comment by ovflowd** [comment]
No, @AugustinMauroy he doesn't need to. He (or anybody else) can do that on a follow-up PR. As this is out of scope of this PR.

File: components/Common/OramaSearchBar/index.module.scss (hunk 1 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.module.scss:1]: Solved. I was using `theme` instead of `resolvedTheme`.

File: components/Common/OramaSearchBar/index.tsx (hunk 2 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx:6]: I think I fixed it now. I let VS code fix it :)

File: components/Common/OramaSearchBar/index.tsx (hunk 3 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx]: You were right. Anyway I removed this after last refactor.

File: components/Common/OramaSearchBar/index.tsx (hunk 4 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx]: Fixed!

File: components/Common/OramaSearchBar/index.tsx (hunk 5 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx:19]: You were right. It is now after component refactoring :)

File: components/Common/OramaSearchBar/index.tsx (hunk 6 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx]: Fixed.

File: components/Common/OramaSearchBar/index.tsx (hunk 7 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx]: I was already using `useTheme` but in the parent component for another stupid reason I'm to embarassed to mention now.
Anyway this is fixed now. :)

File: components/Common/OramaSearchBar/index.tsx
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx:1]: Done. What about now?

File: components/Common/OramaSearchBar/index.tsx (hunk 8 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx]: See above :)

File: components/Common/SearchBar/index.module.scss (hunk 9 above)
Comment (ShogunPanda) [reply components/Common/SearchBar/index.module.scss:128]: I checked and I couldn't find anything.

Instead I found [this](https://github.com/nodejs/nodejs.org/blob/main/styles/_base.scss#L45) which lead me to this implementation.
And in the code (at least in storybook) I see `[data-theme]`. 
Am I missing anything?

**Comment by ovflowd** [comment]
FYI I still need to review the remaining code once I get some time 👀

**Comment by ShogunPanda** [comment]
No worries. Let me know if I have to do anything elsee

File: components/Common/SearchBar/index.module.scss (hunk 9 above)
Comment (shanpriyan) [reply components/Common/SearchBar/index.module.scss:128]: @ovflowd we're currently using `data-theme` for theming, since the `next-theme` introduction.

File: components/Common/OramaSearchBar/index.tsx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import SearchBar from '../SearchBar';
+
+import type { FC } from 'react';
+import type { DocumentMapper, OramaSearchBarProps } from './search';
+import type { SearchFunction } from '../../../types';
+
+const Footer: FC = () => {
+  const { resolvedTheme: theme } = useTheme();
+
+  const isDark = useMemo(() => theme === 'dark', [theme]);
```
Comment (shanpriyan) [inline components/Common/OramaSearchBar/index.tsx]: No need of useMemo here, since there is no heavy computation stuff
```suggestion
  const isDark = theme === 'dark';
```

File: components/Common/OramaSearchBar/index.tsx [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    return documentMapper ?? identityDocumentMapper;
+  }, [documentMapper]);
+
+  // Create a setup function to deserialize the database
+  const setup = useCallback(async () => {
+    const database = await create({ schema });
+    await load(database, index);
+
+    setSearchFunction(() => performSearch.bind(null, database, mapper));
+  }, [schema, index, mapper, setSearchFunction]);
```
Comment (shanpriyan) [inline components/Common/OramaSearchBar/index.tsx]: I think we can use the `documentMapper` directly inside the  setup function

```suggestion
  // Create a setup function to deserialize the database
  const setup = useCallback(async () => {
    const database = await create({ schema });
    await load(database, index);

    setSearchFunction(() => performSearch.bind(null, database, documentMapper ?? identityDocumentMapper));
  }, [schema, index, documentMapper, setSearchFunction]);
```

File: components/Common/SearchBar/index.tsx [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+    setExpanded(false);
+  }, [setQuery, setExpanded]);
+
+  const selectItem = useCallback(
+    (index: number) => {
+      if (!listRef.current) {
+        return;
+      }
+
+      console.log(index);
```
Comment (shanpriyan) [inline components/Common/SearchBar/index.tsx]:
```suggestion
```

File: components/Common/SearchBar/index.tsx [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+}: SearchBarProps) => {
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const searchInputRef = useRef<HTMLInputElement | null>(null);
+  const listRef = useRef<HTMLUListElement | null>(null);
+  const activeResultRef = useRef<number>(-1);
+
+  const [isExpanded, setExpanded] = useState(false);
+  const [query, setQuery] = useState('');
+  const [results, setResults] = useState<SearchResult[]>([]);
+  const isEmpty = useMemo(() => !results || results.length === 0, [results]);
```
Comment (shanpriyan) [inline components/Common/SearchBar/index.tsx]:
```suggestion
  const isEmpty = !results || results.length === 0;
```

File: components/Common/SearchBar/index.tsx [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    height: '3em',
+  },
+};
+
+const SearchBar: FC<SearchBarProps> = ({
+  setup,
+  search,
+  footer: Footer,
+}: SearchBarProps) => {
+  const [isLoading, setIsLoading] = useState<boolean>(true);
```
Comment (shanpriyan) [inline components/Common/SearchBar/index.tsx]:
```suggestion
  const [isLoading, setIsLoading] = useState(true);
```

File: components/Common/SearchBar/index.tsx [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+  // Focus the field when expanding the bar
+  useEffect(() => {
+    if (isExpanded) {
+      searchInputRef.current?.focus();
+    }
+  }, [isExpanded, searchInputRef]);
+
+  // Clear results when the query is emptied
+  useEffect(() => {
+    if (!query && results.length) {
```
Comment (shanpriyan) [inline components/Common/SearchBar/index.tsx]: Can we do this in `onQueryChangeHandler` and remove this useEffect 🤔 ?

This useEffect will result in another render cycle after query has been updated

```js
if(!e.target.value) setResults([])
```

File: i18n/locales/en.json [hunk 16]
```diff
@@ -32,6 +32,11 @@
   "components.downloadReleasesTable.docs": "Docs",
   "components.header.buttons.toggleLanguage": "Toggle Language",
   "components.header.buttons.toggleDarkMode": "Toggle dark/light mode",
+  "components.searchBar.placeholder": "Search",
```
Comment (shanpriyan) [inline i18n/locales/en.json]: Please add the  new keys at the bottom of the file

https://github.com/nodejs/nodejs.org/blob/7a4e29c8090caa8980f51a5676731ea43888f729/TRANSLATION.md?plain=1#L54

File: package.json [hunk 17]
```diff
@@ -36,8 +36,10 @@
   },
   "dependencies": {
     "@mdx-js/react": "^2.3.0",
+    "@mui/icons-material": "^5.11.16",
```
Comment (shanpriyan) [inline package.json]: We are currently using `react-icons` for icons package, update the icons from mui to react-icons

**Review by shanpriyan — CHANGES_REQUESTED** [review]

File: components/Common/OramaSearchBar/index.tsx (hunk 10 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx]: 👍

File: components/Common/OramaSearchBar/index.tsx (hunk 11 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx]: 👍

File: components/Common/SearchBar/index.tsx (hunk 14 above)
Comment (ShogunPanda) [reply components/Common/SearchBar/index.tsx]: 👍

File: i18n/locales/en.json (hunk 16 above)
Comment (ShogunPanda) [reply i18n/locales/en.json]: 👍

File: package.json (hunk 17 above)
Comment (ShogunPanda) [reply package.json]: 👍

**Comment by ovflowd** [comment]
FYI gonna finally review the remaining of the PR this week.

**Comment by ovflowd** [comment]
Relates to https://github.com/nodejs/nodejs.org/issues/5204

File: components/Common/OramaSearchBar/OramaLogo.tsx [hunk 18]
```diff
@@ -0,0 +1,731 @@
...
+      x="0px"
+      y="0px"
+      viewBox="0 0 4207.11 1155.11"
+      className={className}
+      style={{ enableBackground: 'new 0 0 4207.11 1155.11' } as any}
+      xmlSpace="preserve"
+    >
+      <style
+        type="text/css"
+        dangerouslySetInnerHTML={{ __html: logoDarkStyle }}
```
Comment (Harkunwar) [inline components/Common/OramaSearchBar/OramaLogo.tsx]: Could we move the styles to a proper css file so we don't have to use `dangerouslySetInnerHTML`

File: components/Common/SearchBar/index.tsx [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+      e.preventDefault();
+
+      // Clear results when the query is emptied
+      if (!e.target.value) {
+        setResults([]);
+      }
+
+      setQuery(e.target.value);
+    },
+    [setQuery, setResults]
```
Comment (Harkunwar) [inline components/Common/SearchBar/index.tsx:99]:
```suggestion
    []
```
We don't need to add these as any setter resulting from `useState` will never change.

File: components/Common/OramaSearchBar/index.tsx [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+const OramaSearchBar: FC<OramaSearchBarProps> = ({
+  schema,
+  index,
+  documentMapper,
+}: OramaSearchBarProps) => {
+  const [searchFunction, setSearchFunction] = useState<SearchFunction>();
+
+  // Create a setup function to deserialize the database
+  const setup = useCallback(async () => {
```
Comment (Harkunwar) [inline components/Common/OramaSearchBar/index.tsx:48]: Is there any reason why we're passing the setup function to the child instead of just  doing the setup here. It feels more complicated.

File: components/Common/SearchBar/index.tsx [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+    (index: number) => {
+      if (!listRef.current) {
+        return;
+      }
+
+      const el = listRef.current.children[index]!;
+      el.querySelector('a')?.focus();
+      el.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
+    },
+    [listRef]
```
Comment (Harkunwar) [inline components/Common/SearchBar/index.tsx:75]: Refs never change and don't need to be in deps array.

File: components/Common/SearchBar/index.tsx [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+}: SearchBarProps) => {
+  const [isLoading, setIsLoading] = useState(true);
+  const searchInputRef = useRef<HTMLInputElement | null>(null);
+  const listRef = useRef<HTMLUListElement | null>(null);
+  const activeResultRef = useRef<number>(-1);
+
+  const [isExpanded, setExpanded] = useState(false);
+  const [query, setQuery] = useState('');
+  const [results, setResults] = useState<SearchResult[]>([]);
+  const isEmpty = !results || results.length === 0;
```
Comment (Harkunwar) [inline components/Common/SearchBar/index.tsx:46]: I feel like this Component is doing to many things. The interactions  between a lot of the state objects is making it complicated. Can we try to use maybe `useReducer` to see if we can clean things up here?

File: components/Common/OramaSearchBar/index.tsx (hunk 20 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx:48]: I designed `SearchBar` to be backend agnostic, therefore it accepts a `setup` function to prepare (and thus not allow searching before setup has completed.)
If we do as you suggested, then `useCallback` will become a `useEffect`, and then I'll have to change SearchBar to accept a boolean for `isLoading`. I see little benefit here as the setup part is separated from the component anyway.

File: components/Common/SearchBar/index.tsx (hunk 19 above)
Comment (ShogunPanda) [reply components/Common/SearchBar/index.tsx:99]: Here and below: while I agree on that, I tend to pass all values (including immutable one) to useCallback/useMemo in order to have a quick view of what is used in the function. I don't think it affect performance or are there any cons in doing this I'm not aware of?

File: components/Common/SearchBar/index.tsx (hunk 21 above)
Comment (ShogunPanda) [reply components/Common/SearchBar/index.tsx:75]: See https://github.com/nodejs/nodejs.org/pull/5419#discussion_r1251636993

File: components/Common/SearchBar/index.tsx (hunk 22 above)
Comment (ShogunPanda) [reply components/Common/SearchBar/index.tsx:46]: I'm -1 on this.

`query` and `results` are basically changed at the same time.
`expanded` is triggered independently.
The only other state is `isLoading` which is only changed once.

I feel adding adding a reducer would be slightly overengineered.

File: components/Common/OramaSearchBar/OramaLogo.tsx (hunk 18 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/OramaLogo.tsx]: I tried. And I discover none of the styles were actually used. Lol.
I've removed everything.

File: components/Common/OramaSearchBar/index.tsx
Comment (ovflowd) [reply components/Common/OramaSearchBar/index.tsx:1]: Let me re-review 👀

**Comment by ovflowd** [comment]
And here I go reviewing this, sorry for the super significant delay, @ShogunPanda!

File: components/Common/OramaSearchBar/index.module.scss [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+  align-items: center;
+  column-gap: 1ch;
+  display: flex;
+  justify-content: end;
+  margin: 5px;
+
+  .oramaSearchBarFooterLabel {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.module.scss:8]: You can use `&_` statements here, so you don't need to add the whole prefix again. (Actually in all the child Sass styles here)

Eg.:

```
.oramaSearchBarFooter {

  &_label {

  }

}
```

File: components/Common/OramaSearchBar/index.tsx [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+import { OramaLogoDark, OramaLogoLight } from './OramaLogo';
+import styles from './index.module.scss';
+import { identityDocumentMapper, performSearch } from './search';
+import SearchBar from '../SearchBar';
+
+import type { FC } from 'react';
+import type { OramaSearchBarProps } from './search';
+import type { SearchFunction } from '../../../types';
+
+const Footer: FC = () => {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx:14]: Can this become a separate component `OramaSearchBarFooter.tsx`? Just because our approach here is to have one top-level Component per file.

File: components/Common/OramaSearchBar/index.tsx [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+      <span className={styles.oramaSearchBarFooterLabel}>
+        <FormattedMessage id="components.oramaSearchBar.search.poweredBy" />
+      </span>
+      <a
+        href="http://oramajs.io"
+        target="_blank"
+        rel="noopener noreferrer"
+        className={styles.oramaSearchBarFooterLink}
+      >
+        {isDark ? (
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx:30]: Maybe we can simplify this to:

```tsx

const OramaLogo = theme === 'dark' ? OramaLogoDark : OramaLogoLight;

...

<OramaLogo className={styles.oramaSearchBarFooterLogo} />
```

File: components/Common/OramaSearchBar/index.tsx [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+    setSearchFunction(() =>
+      performSearch.bind(
+        null,
+        database,
+        documentMapper ?? identityDocumentMapper
+      )
+    );
+  }, [schema, index, documentMapper, setSearchFunction]);
+
+  return <SearchBar setup={setup} search={searchFunction} footer={Footer} />;
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/index.tsx:61]: I assume that the `SearchBar` awaits for the setup function?

File: components/Common/OramaSearchBar/search.ts [hunk 27]
```diff
@@ -0,0 +1,26 @@
+import { search } from '@orama/orama';
+
+import type { Document, Orama, RawData, Schema } from '@orama/orama';
+import type { SearchResult } from '../../../types';
+
+export type DocumentMapper = (doc: Document) => SearchResult;
+
+export type OramaSearchBarProps = {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/search.ts:8]: I believe types should be solely defined on a `types.ts` file just to separate type definitions from actual `search` logic

File: components/Common/OramaSearchBar/search.ts [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+export const identityDocumentMapper = (doc: Document): SearchResult => {
+  return doc as unknown as SearchResult;
+};
+
+export const performSearch = async (
+  database: Orama,
+  mapper: DocumentMapper,
+  term: string
+): Promise<SearchResult[]> => {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/search.ts:22]: I feel that this could be transformed into a promise. Could we do that?

File: components/Common/OramaSearchBar/search.ts [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+
+export type DocumentMapper = (doc: Document) => SearchResult;
+
+export type OramaSearchBarProps = {
+  schema: Schema;
+  index: RawData;
+  documentMapper?: DocumentMapper;
+};
+
+export const identityDocumentMapper = (doc: Document): SearchResult => {
```
Comment (ovflowd) [inline components/Common/OramaSearchBar/search.ts:14]: Any reason why this function exists? It seems just to be casting one type to another.

File: components/Common/SearchBar/index.tsx
Comment (ovflowd) [inline components/Common/SearchBar/index.tsx:1]: I'm sadly negatively impressed and shocked with the pure size of this Component.

It is so freaking complex. We should discard this component and do something from scratch.

We don't need to keyboard listeners, handlers and all these things now. We can iterate later.

I know this Component is part of the migration, but since it is so complex, it makes it harder to know what's being migrated and what is not.

If I might suggest, let's discard this component and its styles and feel free to make a new one that allows you to type and render the results.

It would be nice if styles, at least of the search container, could be similar, but the content itself, how it collapses/expands, and how it renders the results up to you.

I'm afraid I have to disagree with a Component with many callbacks, effects, conditions and more. Even the `useEffect`'s super complicated.

For example, the rendering of the list (https://github.com/nodejs/nodejs.org/pull/5419/files#diff-cc0fb6b25cb6dad14226e6daa7b1819a859f6c9e5aa102d9557c52b2db485f67R292) is highly complex; we have two effects (https://github.com/nodejs/nodejs.org/pull/5419/files#diff-cc0fb6b25cb6dad14226e6daa7b1819a859f6c9e5aa102d9557c52b2db485f67R157, https://github.com/nodejs/nodejs.org/pull/5419/files#diff-cc0fb6b25cb6dad14226e6daa7b1819a859f6c9e5aa102d9557c52b2db485f67R173) that are also super complex.

I want also to avoid mixing async/await logic on effects, etc., as much as possible. Otherwise, we're creating way too many side-effects.

TL;DR I will find it hard to approve this PR with this Component as it's way too complex. It will on-the-go scare anyone to modify/contribute to this component and it is massively hard to read and understand what it is doing 😅

File: hooks/useKeypress.ts
Comment (ovflowd) [inline hooks/useKeypress.ts:1]: Let's remove this hook from now. It is complex, and the code quality is arguable.

We can think about accessibility on a second pass for this PR. Let's get the basics done, and then we think about invoking the search bar with keystrokes or moving through entries with keystrokes.

File: components/Common/SearchBar/index.tsx
Comment (ovflowd) [reply components/Common/SearchBar/index.tsx:1]: I would also mention that the SearchBar component should probably be broken down in 3 components:

- Search Container
  - This is the actual wrapper where you type your query
- Search Results
  - The container that renders conditionally when there are results
- Search Result / Search Entry
  - The individual element for each result can have different styles

I would also say that SearchResults would have the reference of what is the current highlighted result and handle the keystrokes from going up and down.

And Search Container would handle the keystrokes for expanding/collapsing the search container.

+ on that, I would also say we should make our key-stroke hook very generic once we implement it. But again, all keystroke logic can be done on a follow-up PR.

File: types/search.ts [hunk 30]
```diff
@@ -0,0 +1,12 @@
+export type SearchResult = {
```
Comment (ovflowd) [inline types/search.ts:1]: Usually, on the `types` folder we use interfaces whenever possible. So this one should be an interface.

**Review by ovflowd — CHANGES_REQUESTED** [review]
Overall fantastic work done here, @ShogunPanda.

My review has some minor comments and significant concerns, so I'm sadly requesting changes. Mostly are tied to the SearchBar component.

Thank you so much for your effort here! I hope we're able to address these review concerns so we can move forward 💪

File: components/Common/OramaSearchBar/index.module.scss (hunk 23 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.module.scss:8]: Cool, I'll update this

File: components/Common/OramaSearchBar/index.tsx (hunk 24 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx:14]: I guess so, it's pretty self contained. Will move it soon.

File: components/Common/OramaSearchBar/index.tsx (hunk 25 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx:30]: 👍

File: components/Common/OramaSearchBar/index.tsx (hunk 26 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/index.tsx:61]: Yes, see on line 157.

File: components/Common/OramaSearchBar/search.ts (hunk 27 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/search.ts:8]: Not an issue. Will move them.

File: components/Common/OramaSearchBar/search.ts (hunk 28 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/search.ts:22]: Ehm... it is already a promise. What do you mean exactly?

File: components/Common/OramaSearchBar/search.ts (hunk 29 above)
Comment (ShogunPanda) [reply components/Common/OramaSearchBar/search.ts:14]: The default one is technically a no-op.

Similarly to `SearchBar`, I designed those functions to be use-agnostic. 
Depending on when you're rendering results, you might need to map Orama results to something else. I left room for a simple mapping function to be passed, rather than implementing this inside `useEffect` or similar.

File: components/Common/OramaSearchBar/search.ts (hunk 28 above)
Comment (ovflowd) [reply components/Common/OramaSearchBar/search.ts:22]: You’re mixing async with Promises. I’m referring to use a pure Promise approach for this function

Comment (ShogunPanda) [reply components/Common/OramaSearchBar/search.ts:22]: Ehm... `Promise` it's just the return type. Are looking for something like this?

```
export const performSearch = (
  database: Orama,
  mapper: DocumentMapper,
  term: string
): Promise<SearchResult[]> => {
  return search(database, { term }).then(results => results.hits.map(h => mapper(h.document)));
};
```

File: components/Common/SearchBar/index.tsx
Comment (ShogunPanda) [reply components/Common/SearchBar/index.tsx:1]: Ok, I'll try to split components as you suggested.
Will follow up on this I hope within this week.

File: components/Common/OramaSearchBar/search.ts (hunk 28 above)
Comment (ovflowd) [reply components/Common/OramaSearchBar/search.ts:22]: Yup, that’s what I meant. I also guess we don’t need a return statement as this is an one liner.

Yeah, I meant native promises rather than async/await.

File: components/Common/SearchBar/index.tsx
Comment (ovflowd) [reply components/Common/SearchBar/index.tsx:1]: Yeah I think by doing that, and simplifying the logic and removing everything that is not necessary we make this whole solution way easier.

Comment (ovflowd) [reply components/Common/SearchBar/index.tsx:1]: But let’s also remove unnecessary callbacks, effects and try to simplify them as much as possible. I haven’t gone through the whole component but I guess there’s room for improvements there.

**Comment by ovflowd** [comment]
Hey @ShogunPanda have you had some time to follow-up on this? Do you need any assistance! Thanks :3

**Comment by ShogunPanda** [comment]
Nope, I don't need any help at the moment.
I plan to to work on this next week. I hope to have some updates soon!

**Review by tuanggo — COMMENTED** [review]
open

**Comment by ShogunPanda** [comment]
I mistakenly deleted my fork. Will send a new PR (which fixes all the suggestions outlined here) soon. Sorry for the mess.
